src/main/java/org/spdx/utility/compare/TemplateRegexGenerator.java

```diff
@@ -0,0 +1,322 @@
+					"\\s*";
+		}
+	}
+	
+	static class RegexPattern extends RegexElement {
+		private String pattern;
+		
+		public RegexPattern(String pattern) {
+			this.pattern = pattern;
+		}
+		
+		public String getPattern() {
+			return pattern;
+		}
+		
+		@Override
+		public String toString() {
+			return "(" + pattern + ")"; // We always treat the pattern as a group
+		}
+
+		/**
+		 * @param pattern pattern to set
+		 */
+		public void setPattern(String pattern) {
+			this.pattern = pattern;
+		}
+	}
+
+	private String template;
+	
+	/**
+	 * Top level regex
+	 */
+	private RegexList regexPatternList = new RegexList();
+	
+	
+	private int optionalNestLevel = 0;
+	
+	private List<OptionalRegexGroup> optionalGroups = new ArrayList<>();
+
+	/**
+	 * Generates regular expressions from a license or exception template
+	 * @throws SpdxCompareException 
+	 */
+	public TemplateRegexGenerator(String template) throws SpdxCompareException {
+		this.template = template;
+		parseTemplate();
+	}
+	
+	/**
+	 * Parses the template generating the regular expression
+	 * @throws SpdxCompareException 
+	 */
+	private void parseTemplate() throws SpdxCompareException {
+		try {
+			SpdxLicenseTemplateHelper.parseTemplate(LicenseCompareHelper.removeCommentChars(template), this);
+		} catch (LicenseTemplateRuleException e) {
+			throw new SpdxCompareException("Invalid template rule found during filter: "+e.getMessage(),e);
+		} catch (LicenseParserException e) {
+			throw new SpdxCompareException("Invalid template found during filter: "+e.getMessage(),e);
+		}
+	}
+	
+	/**
+	 * @return the complete regular expression for the template
+	 */
+	public String getCompleteRegex() {
+		return regexPatternList.toString();
+	}
+	
+	/**
+	 * @param wordLimit number of non optional words to include in the pattern
+	 * @return a regex to match the start of the license per the template
+	 */
+	public String getStartRegex(int wordLimit) {
+		RegexList result = new RegexList();
+		int index = 0;
+		int numWords = 0;
+		List<RegexElement> elementList = regexPatternList.getElements();
+		while (index < elementList.size() && numWords <= wordLimit) {
+			RegexElement element = elementList.get(index++);
+			result.addElement(element);
+			if (element instanceof RegexToken) {
+				numWords++;
+			}
+		}
+		// Need to check if the string starts with a greedy regex and change to non-greedy
+		if (result.getElements().size() > 0) {
+			RegexElement firstElement = result.getElements().get(0);
+			if (firstElement instanceof RegexPattern) {
+				String pattern = ((RegexPattern)firstElement).pattern;
+				if (!pattern.startsWith(".?") && pattern.startsWith(".")) {
+					((RegexPattern)firstElement).setPattern(".?" + pattern.substring(1));
+				}
+			}
+		}
+		return result.toString();
+	}
+	
+	/**
+	 * @param wordLimit number of non optional words to include in the pattern
+	 * @return a regex to match the end of the license per the template
+	 */
+	public String getEndRegex(int wordLimit) {
```

Comment:
Might this be private?  Or are there cases where callers (e.g. `LicenseCompareHelper`) actually need the end regex independently of the entire thing?